hw/ip/otbn/dv/rig/rig/gens/bad_branch.py

```diff
@@ -0,0 +1,124 @@
+
+    '''
+
+    ends_program = True
+
+    def __init__(self, cfg: Config, insns_file: InsnsFile) -> None:
+        super().__init__()
+
+        self.insns = []
+        self.weights = []
+
+        self.beq = self._get_named_insn(insns_file, 'beq')
+        self.bne = self._get_named_insn(insns_file, 'bne')
+
+        # beq and bne expect operands: grs1, grs2, offset
+        for insn in [self.beq, self.bne]:
+            if not (len(insn.operands) == 3 and
+                    isinstance(insn.operands[0].op_type, RegOperandType) and
+                    insn.operands[0].op_type.reg_type == 'gpr' and
+                    not insn.operands[0].op_type.is_dest() and
+                    isinstance(insn.operands[1].op_type, RegOperandType) and
+                    insn.operands[1].op_type.reg_type == 'gpr' and
+                    not insn.operands[1].op_type.is_dest() and
+                    isinstance(insn.operands[2].op_type, ImmOperandType) and
+                    insn.operands[2].op_type.signed):
+                raise RuntimeError('{} instruction from instructions file is not '
+                                   'the shape expected by the Branch generator.'
+                                   .format(insn.mnemonic))
+
+        self.imm_op_type = self.bne.operands[2].op_type
+
+        for insn in insns_file.insns:
+            if (insn.mnemonic in ['beq', 'bne']):
+                weight = cfg.insn_weights.get(insn.mnemonic)
+                if weight > 0:
+                    self.weights.append(weight)
+                    self.insns.append(insn)
+
+        # Check that at least one instruction has a positive weight
+        assert len(self.insns) == len(self.weights)
+        if not self.weights:
+            self.disabled = True
+
+    def gen(self,
+            cont: GenCont,
+            model: Model,
+            program: Program) -> Optional[GenRet]:
+
+        # Max/Min Offsets for BXX (-2048 * 2, 2047 * 2)
+        # We cannot use get_op_val_rng methods because it is clever enough
+        # to calculate range with regards to PC.
+        max_offset = 4094
+        min_offset = -4096
+
+        # Get known registers
+        known_regs = model.regs_with_known_vals('gpr')
+
+        # Pick a random register among known registers. We always have x0.
+        # So it should never fail.
+        idx, value = random.choice(known_regs)
+        assert idx is not None
+
+        # Get the chosen base register index as grs1 and grs2 operand. This is
+        # because we want to branch to the faulty addresses with this snippet.
+        op_val_grs1 = idx
+        assert op_val_grs1 is not None
+
+        # We need to pick an out of bounds offset
+        imm_val = random.randrange(min_offset, max_offset, 2)
+        while not ((imm_val + model.pc < 0) or
+                   (imm_val + model.pc > program.dmem_size)):
+            imm_val = random.randrange(min_offset, max_offset, 2)
```

Comment:
Assuming we've re-jigged things so that we're picking the absolute destination instead of an offset, couldn't you do this by intersecting the representable range with `(-∞, 0) ∩ (dmem_size, ∞)`? I think this would avoid needing the while loop.